### ISSUE TITLE
Warn for ignored assignments

### ISSUE THREAD

**Issue by Keisial** [koalaman/shellcheck#2300]
Checklist: I checked both https://www.shellcheck.net/ and a local compile of master (da7b2821). I searched through issues and didn't find any existing issue for this.

Consider the following snippet:
```sh
#!/bin/sh
GROUPS="cabal"

if [ "$GROUPS" = "cabal" ]; then
  echo "There is a cabal!"
else
  echo "There is no cabal"
fi
```
Now, _depending on the shell_, it will print an echo or the other. ksh and dash will print that there is a cabal, whereas on bash the assignment is silently ignored (even when running as `/bin/sh`) and echo the second one.

#### Here's what shellcheck currently says:
Currently, shellcheck gives out (regardless of the dialect) a `SC2178: Variable was used as an array but is now assigned a string.` (which isn't completely correct, since it's only an array in bash).

shellcheck will be completely silent for
```bash
#!/bin/bash
GROUPS=("cabal")

if [ "${GROUPS[0]}" = "cabal" ]; then
  echo "There is a cabal!"
else
  echo "There is no cabal"
fi
```

This has been the subject of the recent [bug-bash thread GROUPS](https://lists.gnu.org/archive/html/bug-bash/2021-08/msg00013.html) (a real-world script was using `GROUPS` as variable name, and it broke when changing from a server where `/bin/sh` was provided by `dash` to one where it was provided by `bash`)

#### Here's what I wanted or expected to see:

I think shellcheck ought output a new warning:  `SCxyz: Assignments to a variable named GROUPS are ignored by bash`

One might additionally check the dialect, or detect first if there was a prior `unset GROUPS` (which would make it lose its magic), but I think it's fine to always warn on assigning to such variable. It's misleading enough to completely avoid that name as a user variable.

bash variables for which assignments have no effect and should thus trigger this new warning are:
`BASHPID`, `EPOCHREALTIME`, `EPOCHSECONDS`, `FUNCNAME`, `GROUPS`.
